LIPS/lip-2.md

```diff
@@ -0,0 +1,248 @@
+  reverts the transaction with `ALLOWED_BEACON_BALANCE_INCREASE` code.
+* if there is a loss, calculates relative decrease and compares it with the lower bound. If was
+  below, reverts the transaction with `ALLOWED_BEACON_BALANCE_DECREASE` code.
+
+
+## Callback function to be invoked on report pushes.
+
+To provide the external contract with updates on report pushes (every time the quorum is reached
+among oracle daemons data), we provide the following setter and getter functions.
+
+    function setQuorumCallback(address _addr) external auth(SET_QUORUM_CALLBACK)
+
+    function getQuorumCallback() public view returns(address)
+
+And when the callback is set, the following function will be invoked on every report push.
+
+    interface IQuorumCallback {
+        function processLidoOracleReport(
+            uint256 _postTotalPooledEther,
+            uint256 _preTotalPooledEther,
+            uint256 _timeElapsed) external;
+    }
+
+The arguments provided are the same as described in sections above.
+
+The following contract storage variables are used to keep the information.
+
+    bytes32 internal constant QUORUM_CALLBACK_POSITION = keccak256("lido.LidoOracle.quorumCallback");
+
+
+## Add events to cover all states change
+
+The goal was to cover every state change and have access to every storage variable. This is why we
+added the following events.
+
+    event BeaconSpecSet(
+        uint64 epochsPerFrame,
+        uint64 slotsPerEpoch,
+        uint64 secondsPerSlot,
+        uint64 genesisTime
+    );
+
+Reports beacon specification update by governance.
+
+    event ReportableEpochIdUpdated(uint256 epochId)
+    
+Reports the new epoch that the contract is ready to accept from oracles. This happens as a result of
+either a successful quorum or when some oracle reported later epoch.
+    
+    event BeaconReported(
+        uint256 epochId,
+        uint128 beaconBalance,
+        uint128 beaconValidators,
+        address caller
+    )
+    
+Reports the data that the oracle pushed to the contract with `reportBeacon` call.
+    
+    event PostTotalShares(
+         uint256 postTotalPooledEther,
+         uint256 preTotalPooledEther,
+         uint256 timeElapsed,
+         uint256 totalShares)
+
+Reports statistics data when the quorum happened and the resulting report was pushed to Lido. Here,
+`totalShares` is grabbed from the Lido contract, see [Add calculation of staker rewards APR][3]
+section above for other arguments.
+         
+    event AllowedBeaconBalanceAnnualRelativeIncreaseSet(uint256 value)
+    event AllowedBeaconBalanceRelativeDecreaseSet(uint256 value)
+
+Reports the updates of the threshold limits by the governance. 
+See [Sanity checks the oracles reports by configurable values][4] section above for details.
+
+    event QuorumCallbackSet(address callback);
+
+Reports the updates of the quorum callback, [Callback function to be invoked on report pushes][5].
+
```

Comment:
```suggestion

It can be reasonably argued that these events and getters belong to Lido app rather than an oracle app. We would agree to that in a vacuum, but the reality of the situation is that the ecosystem needs these events and getters to build upon, and we're not upgrading the core contract anytime soon. So we're taking a workaround and putting some code that best belongs elsewhere in the oracle.
```

Reply:
Good, added